### DISCUSSION TITLE
Best Practices for Inserting Large Number of Rows

### DISCUSSION THREAD

**Discussion by MaxwellGover** [supabase/supabase#11349] [Questions] [answered]
Hello. I don't have much db experience. My question is regarding best practices for inserting large amounts of rows to my supabase db. The amount of rows is `55515` at the moment and takes easily over an hour. I actually don't know the real number because I've never let the process complete. Currently I'm inserting all of the data all at once like so

```
const { data, error } = await supabase
    .from('player_game_logs')
    .insert(dbData)
```

Would it be faster to map over each item and try to save it or split my array into a bunch of different segments and save it that way? 

Thanks

**Comment by kiwicopple** [comment, marked as answer]
See "Bulk Loading Data" here: https://supabase.com/docs/guides/database/tables#bulk-data-loading

TLDR - you should use the postgres `COPY` command (it's _much_ faster than using a REST API).

**Reply by Johnrobmiller** [reply]
not going to lie, was really hoping for a frontend function to handle. I can't exactly run the postgres CLI on the front end

**Reply by kiwicopple** [reply]
ah sorry, if you simply need to insert a lot of rows from your own frontend, use the upsert method:

https://supabase.com/docs/reference/javascript/upsert?example=bulk-upsert-your-data

<img width="634" alt="image" src="https://github.com/supabase/supabase/assets/10214025/96e6f16a-5833-4cb9-a23f-06adcb70239f">

**Comment by johannbuscail** [comment]
Any update on this ?

**Comment by gazreyn** [comment]
I suppose using an edge function or database function would be the best route here? A single payload from the user which then gets processed and inserted in to the DB as a single transaction.

**Comment by vargasjona** [comment]
What worked for me was inserting in a batch of 500 with this code in typescript in the past I inserted 23000 in 36 seconds. Split in groups of 500 items showed me the best insert time in my case.

```ts

export function chunkArray<T>(array: T[], size: number): T[][] {
	const chunks: T[][] = [];
	for (let i = 0; i < array.length; i += size) {
		chunks.push(array.slice(i, i + size));
	}
	return chunks;
}

.....................................
// users comes from a Excel sheet and was converted into an array of objects
const chunkedUsers = chunkArray(users, 500);

for (let i = 0; i < chunkedUsers.length; i++) {
    const { error } = await supabaseClient
        .from('user')
        .insert(chunkedUsers[i]);

    if (error) {
        logger.error(`Insert users error in iteration ${i}`, error)
    }
}

```

**Reply by guidesify** [reply]
Just did this as well. Managed to upsert 2 million rows in the matter of minutes. Do note that this will cause a spike in usage though.

**Comment by zoltlabs** [comment]
Here's a Python script I wrote to do this fast. Uploaded millions of rows in a few minutes. You can also just run it in a Jupyter notebook if you prefer that.

```
import csv
import json
from datetime import datetime, timezone

# Configuration
FILE_PATH = 'path/to/your/file.csv'
TABLE_NAME = 'your_table_name'
BATCH_SIZE = 500
START_PROCESSING_AT_ROW = 0  # Set to 0 to process all rows
csv.field_size_limit(100000000)

def process_row(row):
    """
    Process a row from the CSV file.
    Modify this function to fit your specific data processing needs.
    """
    for key, value in row.items():
        # Convert empty strings to None
        if value == '':
            row[key] = None
        # Convert 'true'/'false' strings to boolean
        elif value.lower() in ['true', 'false']:
            row[key] = value.lower() == 'true'
        # Try to convert to int or float
        else:
            try:
                row[key] = int(value)
            except ValueError:
                try:
                    row[key] = float(value)
                except ValueError:
                    # If it's not a number, try to parse as JSON
                    try:
                        row[key] = json.loads(value)
                    except json.JSONDecodeError:
                        # If it's not JSON, keep it as a string
                        pass

    # Set timestamps
    current_time = datetime.now(timezone.utc).isoformat()
    row['updated_at'] = current_time
    row['created_at'] = current_time

    return row

def upsert_batch(db, table_name, rows):
    """
    Upsert a batch of rows into the specified table.
    Modify this function to match your database API.
    """
    try:
        response = db.table(table_name).upsert(rows).execute()
        if response.data:
            print(f"Successfully upserted batch of {len(rows)} rows.")
        else:
            print(f"Failed to upsert batch of {len(rows)} rows.")
    except Exception as e:
        print(f"Error upserting batch: {str(e)}")

def main(db):
    with open(FILE_PATH, 'r', newline='', encoding='utf-8') as csvfile:
        csv_reader = csv.DictReader(csvfile)
        
        rows_to_upsert = []
        total_rows = 0
        
        for row in csv_reader:
            total_rows += 1
            
            if total_rows <= START_PROCESSING_AT_ROW:
                continue
            
            processed_row = process_row(row)
            rows_to_upsert.append(processed_row)
            
            if len(rows_to_upsert) == BATCH_SIZE:
                upsert_batch(db, TABLE_NAME, rows_to_upsert)
                rows_to_upsert = []  # Clear the list for the next batch
            
            if total_rows % BATCH_SIZE == 0:
                print(f"Processed {total_rows} rows...")
        
        # Upsert any remaining rows
        if rows_to_upsert:
            upsert_batch(db, TABLE_NAME, rows_to_upsert)

    print(f"Finished processing {total_rows} rows from the CSV file and upserting into the database.")

if __name__ == "__main__":
    # Initialize your database connection here
    db = None  # Replace with your Supabase create_client.
    main(db)
```

**Comment by Revadike** [comment]
In case someone is reading this. I looked at the supabase dashboard CSV import function. They convert the csv to json and execute the following query:
```sql
insert into public.table (col1,col2) select col1,col2 from jsonb_populate_recordset(null::public.tags, '[{"col1":"val1","col2":"val2"}]');
```
So, based on that I made my own postgresql function:
```sql
CREATE OR REPLACE FUNCTION bulk_insert(json_data jsonb)

RETURNS void AS $$

BEGIN

    EXECUTE format(

        'INSERT INTO table SELECT * FROM jsonb_populate_recordset(null::table, %L)',

        json_data

    );

END;

$$ LANGUAGE plpgsql;
```
(Use security definer if you want to bypass RLS)

This I can call as RPC function from client, like so:
```js
await supabase.rpc('bulk_insert', {
    json_data: [
        { col1: 'val1', col2: 'val2' },
        { col1: 'val1', col2: 'val2' },
        { col1: 'val1', col2: 'val2' },
        { col1: 'val1', col2: 'val2' },
        { col1: 'val1', col2: 'val2' },
        { col1: 'val1', col2: 'val2' },
        { col1: 'val1', col2: 'val2' },
        { col1: 'val1', col2: 'val2' },
        { col1: 'val1', col2: 'val2' },
  });
```

Works like a charm! Great for Edge functions.

**Reply by dule-kitemetric** [reply]
Thank you @Revadike, I'm going to use your approach since i want to insert data into multiple tables, which should be in a single PostgreSQL function so that it executes script in a transaction


Just my curiousity, is there any way that we can define more specific type for the parameter instead of JSONB?
E.g. something like this
```TS
bulk_insert(data public.my_table[])
```
... or we have to define a TypeScript type/interface for the JSON format and have to ensure to pass matching parameter?

**Reply by Revadike** [reply]
> @Revadike Just my curiousity, is there any way that we can define more specific type for the parameter instead of JSONB? E.g. something like this
> 
> ```ts
> bulk_insert(data public.my_table[])
> ```
> 
> ... or we have to define a TypeScript type/interface for the JSON format and have to ensure to pass matching parameter?

Hmm maybe [pg_jsonschema](https://supabase.com/blog/pg-jsonschema-a-postgres-extension-for-json-validation) is what you are looking for?

**Reply by dule-kitemetric** [reply]
I've just read [pg_jsonschema](https://supabase.com/blog/pg-jsonschema-a-postgres-extension-for-json-validation), it helps validate the JSON data without creating additional database object.

Thank you so much for the suggestion @Revadike. It's really helpful.

**Comment by sebworks2024** [comment]
You can also connect to the database pool directly and use the copy command. It's much [faster](https://dev.to/josethz00/speed-up-your-postgresql-bulk-inserts-with-copy-40pk).

You will need to use the connection properties provided by Supabase and the [pg](https://www.npmjs.com/package/pg) package.
